counter/src/main/java/org/infinispan/counter/impl/strong/AbstractStrongCounter.java

```diff
@@ -86,9 +87,30 @@ public static CompletionStage<Void> removeStrongCounter(Cache<StrongCounterKey,
       return cache.removeAsync(new StrongCounterKey(counterName)).thenApply(CompletableFutures.toNullFunction());
    }
 
-   public final void init() {
+   public final CompletionStage<InternalCounterAdmin> init() {
       registerListener();
-      awaitCounterOperation(readOnlyMap.eval(key, ReadFunction.getInstance()).thenAccept(this::initCounterState));
+      return readOnlyMap.eval(key, ReadFunction.getInstance()).thenAccept(this::initCounterState).thenApply(unused -> this);
+   }
+
+   @Override
+   public StrongCounter asStrongCounter() {
+      return this;
+   }
+
+   @Override
+   public CompletionStage<Void> destroy() {
+      removeListener();
```

Comment:
I must admit I am not sure whose responsibility cleaning up the various listeners and such are. This method just removes from a map, is that sufficient, if you could confirm :+1:

Reply:
It is enough. The map contains the `CounterListener` instances registered by the user code. The clustered listener register in the cache instance is still there :) (and used by other counters)